### PR TITLE
Switch Sparse GP to the QR Decomposition

### PR DESCRIPTION
**Pull request by akleeman**
This PR addresses a numerical stability issue we noticed with the previous approach to the `SparseGaussianProcess`.  In particular the posterior covariance used to be computed as:
```
  C = L^-T R^T R B^-1 L^-1
```
Notice that there is not way to break that formula into a square root form and as a result what was happening is that as the system becomes ill-conditioned not only does the mean computation become unstable (at some point that's unavoidable regardless of approach) but the predicted covariance also because asymmetric.  This would cause steps downstream to fail since subsequent computations using the covariance might fail hard.

To fix this we switched to the QR based approach proposed in,

[Stable and Efficient Gaussian Process Calculations](http://www.jmlr.org/papers/volume10/foster09a/foster09a.pdf)

An outline of the specifics to our approach was included in code comments, but here is a (much) more detailed description:

[Sparse_Gaussian_Process.pdf](https://github.com/swift-nav/albatross/files/4494112/Sparse_Gaussian_Process.pdf)

@pgrgich

### CODE REVIEW

File: include/albatross/src/eigen/serializable_ldlt.hpp [hunk 1]
```diff
@@ -52,20 +53,66 @@ class SerializableLDLT : public LDLT<MatrixXd, Lower> {
...
+  /*
+   * Computes the square root of the diagonal, D^{1/2}
+   */
+  Eigen::DiagonalMatrix<double, Eigen::Dynamic> diagonal_sqrt() const {
+    Eigen::VectorXd thresholded_diag = this->vectorD();
+    for (Eigen::Index i = 0; i < thresholded_diag.size(); ++i) {
+      if (thresholded_diag[i] > 0.) {
+        thresholded_diag[i] = std::sqrt(thresholded_diag[i]);
+      } else {
+        thresholded_diag[i] = 0.;
```
Comment (seth-swiftnav) [inline include/albatross/src/eigen/serializable_ldlt.hpp:81]: Should these functions be asserting the diagonal is nonnegative? Can we get a zero diagonal? It's valid for this function, but a problem everywhere else in LDLT.

File: include/albatross/src/utils/linalg_utils.hpp [hunk 2]
```diff
@@ -15,6 +15,37 @@
 
 namespace albatross {
 
+inline Eigen::MatrixXd
+get_R(const Eigen::ColPivHouseholderQR<Eigen::MatrixXd> &qr) {
+  // Unfortunately the matrixR() method in Eigen's QR decomposition isn't
+  // actually the R matrix, it's tall skinny matrix whose lower trapezoid
+  // contains internal data, only the upper triangular portion is useful
```
Comment (seth-swiftnav) [inline include/albatross/src/utils/linalg_utils.hpp:22]: I guess Eigen never read Clean Code.

File: tests/test_async_utils.cc [hunk 3]
```diff
@@ -83,7 +83,6 @@ TEST(test_async_utils, test_async_apply_map_key_value_function) {
     std::lock_guard<std::mutex> lock(mu);
     sum += x;
     order_processed.push_back(x);
-    std::cout << key << std::endl;
```
Comment (seth-swiftnav) [inline tests/test_async_utils.cc:86]: Sorry.

File: include/albatross/src/eigen/serializable_ldlt.hpp (hunk 1 above)
Comment (akleeman) [reply include/albatross/src/eigen/serializable_ldlt.hpp:81]: While in theory we should never get a negative diagonal for a PSD matrix numerically we can and this logic amounts to zeroing out any of the numerically negative definitely directions and mimics what Eigen does internally:

https://github.com/libigl/eigen/blob/master/Eigen/src/Cholesky/LDLT.h#L578-L584

We actually assert that the diagonal is positive here:

https://github.com/swift-nav/albatross/blob/master/include/albatross/src/eigen/serializable_ldlt.hpp#L39

but I'm not convinced that's the correct thing to do, we should probably just be warning and continuing but unfortunately we don't have logging setup in albatross yet.

File: tests/test_async_utils.cc (hunk 3 above)
Comment (akleeman) [reply tests/test_async_utils.cc:86]: np, looks like you removed it too.

**Comment by pgrgich** [comment]
@akleeman it doesn't look like `Sparse_Gaussian_Process.pdf` is referenced in the comments in the code anywhere. Can that be put in?

**Comment by akleeman** [comment]
@pgrgich I figured I'd update the documentation with a version of the pdf, otherwise (besides adding the pdf directly to the repo) I'm not sure how we'd host a public version of it.  That make sense?

**Comment by pgrgich** [comment]
@akleeman Oh, because `albatross` is publicly available. Hmm, that makes it tricky. I was thinking og putting it in Confluence, but I guess that doesn't work. Ok, sounds good.

**Comment by akleeman** [comment]
https://github.com/swift-nav/albatross/pull/229

**Pull request merged**